src/or/circuituse.c

```diff
@@ -187,35 +190,47 @@ circuit_is_acceptable(const origin_circuit_t *origin_circ,
           /* we don't know the digest; have to compare addr:port */
           if (family < 0 ||
               !tor_addr_eq(&build_state->chosen_exit->addr, &addr) ||
-              build_state->chosen_exit->port != conn->socks_request->port)
+              build_state->chosen_exit->port !=
+              entry_conn->socks_request->port)
             return 0;
         }
       }
     } else {
-      if (conn->want_onehop) {
+      if (entry_conn->want_onehop) {
         /* don't use three-hop circuits -- that could hurt our anonymity. */
         return 0;
       }
     }
     if (origin_circ->prepend_policy && family != -1) {
       int r = compare_tor_addr_to_addr_policy(&addr,
-                                              conn->socks_request->port,
+                                              entry_conn->socks_request->port,
                                               origin_circ->prepend_policy);
       if (r == ADDR_POLICY_REJECTED)
         return 0;
     }
-    if (exitnode && !connection_ap_can_use_exit(conn, exitnode)) {
+    /* If we can't exit on IPv4, check if the node supports exiting on IPv6. */
+    if (!entry_conn->entry_cfg.ipv4_traffic &&
+        conn->purpose == CONN_TYPE_EXIT) {
+      tor_addr_t addr6;
+      tor_addr_make_null(&addr6, AF_INET6);
+      int r = compare_tor_addr_to_node_policy(&addr6,
+                                              entry_conn->socks_request->port,
+                                              exitnode);
+      if (r == ADDR_POLICY_REJECTED)
+        return 0;
```

Comment:
I think all this should go in `connection_ap_can_use_exit()`. But also, how is this different from what there is already in that function? Maybe we don't do the check correctly?